server/__init__.py

```diff
@@ -103,39 +107,54 @@ def report_dirties():
             player_service.clear_dirty()
 
             if len(dirty_queues) > 0:
-                ctx.broadcast_raw(encode_queues(dirty_queues), lambda lobby_conn: lobby_conn.authenticated)
+                await ctx.broadcast_raw(
+                    encode_queues(dirty_queues),
+                    lambda lobby_conn: lobby_conn.authenticated
+                )
 
             if len(dirty_players) > 0:
-                ctx.broadcast_raw(encode_players(dirty_players), lambda lobby_conn: lobby_conn.authenticated)
-
-            # TODO: This spams squillions of messages: we should implement per-connection message
-            # aggregation at the next abstraction layer down :P
+                await ctx.broadcast_raw(
+                    encode_players(dirty_players),
+                    lambda lobby_conn: lobby_conn.authenticated
+                )
+
+            # TODO: This spams squillions of messages: we should implement per-
+            # connection message aggregation at the next abstraction layer down :P
+            tasks = []
             for game in dirty_games:
                 if game.state == GameState.ENDED:
                     games.remove_game(game)
 
                 # So we're going to be broadcasting this to _somebody_...
                 message = encode_dict(game.to_dict())
 
-                # These games shouldn't be broadcast, but instead privately sent to those who are
-                # allowed to see them.
+                # These games shouldn't be broadcast, but instead privately sent
+                # to those who are allowed to see them.
                 if game.visibility == VisibilityState.FRIENDS:
-                    # To see this game, you must have an authenticated connection and be a friend of the host, or the host.
-                    validation_func = lambda lobby_conn: lobby_conn.player.id in game.host.friends or lobby_conn.player == game.host
+                    # To see this game, you must have an authenticated
+                    # connection and be a friend of the host, or the host.
+                    def validation_func(lobby_conn):
+                        return lobby_conn.player.id in game.host.friends or \
+                               lobby_conn.player == game.host
                 else:
-                    validation_func = lambda lobby_conn: lobby_conn.player.id not in game.host.foes
+                    def validation_func(lobby_conn):
+                        return lobby_conn.player.id not in game.host.foes
+
+                tasks.append(ctx.broadcast_raw(
+                    message,
+                    lambda lobby_conn: lobby_conn.authenticated and validation_func(lobby_conn)
+                ))
+
+            await asyncio.gather(*tasks)
```

Comment:
Is it safe to run all these broadcasts in parallel? Is it possible for data they send to get interleaved?

Reply:
Well you have to remember that with asyncio, nothing actually runs in parallel. There is no scheduler that will interrupt a coroutine, so once a coroutine starts, it keeps running until it decides it’s ready to stop. So I don’t imagine that it’s possible for any messages to become corrupted. I would also guess that they’ll likely execute in the order that they’re created, although it wouldn’t be guaranteed.

Reply:
Only chunks of code that don't contain any await calls are guaranteed to be atomic, so if we sent any command in chunks using multiple await calls, then they could get interleaved. It does seem that we're only writing whole commands, though.

Reply:
Yea, I don't think the code does that anywhere. The `QDataStreamProtocol.send_message` function always encodes an entire message and writes it with a single call to `StreamWriter.write`